### PR TITLE
Replace SyntaxHighlighter with Monaco Editor.

### LINKED ISSUE #3832: Large response bodies cause hanging
| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  Bug
| Which Swagger/OpenAPI version?  | 2.0
| Which Swagger-UI version?       | 3.4.0
| How did you install Swagger-UI? | local dev server
| Which browser & version?        | Chrome 61
| Which operating system?         | macOS High Sierra

### Demonstration API definition

Swagger 2 Petstore is sufficient.

### Configuration (browser query string, constructor, config.yaml)

Default.

### Current Behavior

As an example, `GET /pet/findByStatus` in the Petstore currently returns 1.8MB of JSON when all statuses are selected. This is a pretty big chunk of data, but Swagger-UI insists on rendering the _entire_ body, which is neither performant or useful. The main thread of the application is locked for upwards of 20 seconds on my machine.

### Possible Solution

Truncate or refuse to display large textual response bodies.

### Context

`GET /pet/findByStatus` is my favorite operation to use when testing UI enums, so this is something that impedes me pretty regularly.

### PR DESCRIPTION
**Pull request by nojaf**
### Description

This replaces `react-syntax-highlighter` with `@monaco-editor/react`.

### Motivation and Context

Fixes #3832.

The key motivation for us is to see large response bodies all the time.

### How Has This Been Tested?

I've been playing with this manually on my local machine.
The endpoint I'm testing is retrieving 2.8 MB of JSON and it gets instantly rendered in my browser.

### Screenshots (if appropriate):

image

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [x] Dependency changes (any modification to dependencies in `package.json`)
- [?] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [x] are breaking changes to a public API (config options, System API, major UI change, etc).

I'm not quite sure yet what to do with the `syntaxHighlight.theme` setting. These themes were tied to `SyntaxHighlighter` and the Monaco editor has different themes out of the box.
Please let know how we can deal with this.

- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [ ] are not breaking changes.

### Documentation
- [ ] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.

I'm not sure here: please tell me.

- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [ ] All new and existing tests passed.

As I'm new to this, I'm not quite sure what needs to be done here. Again, let me know and I'll look into it.

### CODE REVIEW

**Review by heldersepu — CHANGES_REQUESTED** [review]

**Comment by heldersepu** [comment]
https://raw.githack.com/nojaf/swagger-ui/large-bodies/dist/index.html#/pet/findPetsByStatus
I tested your version  and I'm not getting the same as your image...

File: test/unit/components/highlight-code.jsx [hunk 1]
```diff
@@ -18,13 +18,12 @@ describe("<HighlightCode />", () => {
     expect(wrapper.find("CopyToClipboard").length).toEqual(1)
   })
 
-  it("should render values in a preformatted element", () => {
+  it("should render values in a section element", () => {
     const value = "test text"
     const props = {value: value, getConfigs: fakeGetConfigs}
     const wrapper = mount(<HighlightCode {...props} />)
-    const preTag = wrapper.find("pre")
+    const sectionTag = wrapper.find("section")
```
[the remark this answers is not in the document]
Comment (heldersepu) [reply test/unit/components/highlight-code.jsx:25]: Nevermind ... I just realized this is just a test.

**Comment by heldersepu** [comment]
> https://raw.githack.com/nojaf/swagger-ui/large-bodies/dist/index.html#/pet/findPetsByStatus I tested your version and I'm not getting the same as your image...
> 
> the `<div class="highlight-code">` seem to be just a few pixels of height
> 
I think the issue here might be just old bundle, I added my changes to my branch and the height is fine:
https://raw.githack.com/heldersepu/swagger-ui/test/dist/index.html?docExpansion=None&defaultModelsExpandDepth=None#/pet/findPetsByStatus

**Comment by heldersepu** [comment]
I change the theme to be dark and disabled the minimap to get as close as possible as it was:

Screenshot from 2024-03-23 19-43-59

simple changes:
https://github.com/swagger-api/swagger-ui/commit/45c763181bb968aa180f306cb1f68f7b5f8aebdf

File: test/unit/components/highlight-code.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (nojaf) [reply test/unit/components/highlight-code.jsx:25]: Yes, I believe `react-syntax-highlighter` created a `pre` tag and that was the assertion in the test.

**Comment by nojaf** [comment]
Hi @heldersepu, thank you for the feedback!
I believe I addressed all your remarks.

Any thoughts on what to do with the `syntaxHighlight.theme` setting?

And could I get the workflow approval?
image

**Comment by heldersepu** [comment]
@nojaf rebuild the bundles when you can I still can see the old version:
https://raw.githack.com/nojaf/swagger-ui/large-bodies/dist/index.html#/pet/findPetsByStatus 

would be nice to integrate the `syntaxHighlight.theme` into your changes

I'm not in control of the workflow some one from the team should do that

**Comment by heldersepu** [comment]
@glowcloud & @char0n  👀

**Comment by nojaf** [comment]
Hi @heldersepu, apologies for the delay. I took some time off.
I've rebuilt all bundles.

> would be nice to integrate the syntaxHighlight.theme into your changes

Happy to do this, but could use a pointer. Do we care about the existing values?

**Comment by heldersepu** [comment]
I don't think the existing theme values are compatible with the Monaco editor 
test see what happens if an invalid theme is given, maybe there is a way we can provide a default or validate is one of the allowed themes

**Comment by nojaf** [comment]
So, some e2e tests are failing like:

https://github.com/swagger-api/swagger-ui/blob/3e81a4f897e4be9530f3cc6b76e50570e3e61da4/test/e2e-cypress/e2e/bugs/4943.cy.js#L10-L19

#4943 XML example not rendered correctly with oneOf -- should render oneOf property correctly (failed)

The problem presumably is in the assertion and might not be an actual problem.

To run this: `npm run cy:start` and in a different terminal `npm run cy:run -- --spec "**/4943.cy.js"`.

Could I please get some pointers on how to address this?
I have never seen this Cypress thing up close and it is a tad overwhelming.

**Comment by heldersepu** [comment]
@nojaf don't feel discouraged this PR is great, the syntax highlighter was in need of a major refactor...
The team must be busy with other priorities

**Comment by heldersepu** [comment]
When you can address the:
This branch has conflicts that must be resolved

**Comment by char0n** [comment]
Thanks for all the work on this, but we'll not be incorporating this major architectural change into core SwaggerUI. Due to https://github.com/swagger-api/swagger-ui/pull/9783, this can now be approached using the standard SwaggerUI plugin system. More information in https://github.com/swagger-api/swagger-ui/issues/3832#issuecomment-2047446217

I've also created a wiki page with your MonacoEditor plugin - https://github.com/swagger-api/swagger-ui/wiki/Plugins

**Comment by char0n** [comment]
@nojaf it's experiments like these that push the evolution of SwaggerUI forward. Your PR did directly inspire https://github.com/swagger-api/swagger-ui/pull/9783

**Pull request closed without merging**